misc/format.py

```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/python3
+
+import os, re
+import subprocess as sp
+from pathlib import Path
+from colorama import Fore, Back, Style
+from yapf.yapflib.yapf_api import FormatFile
+from git import Repo
+
+repo_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+
+
+def has_suffix(f, suffixes):
+    for suf in suffixes:
+        if f.endswith('.' + suf):
+            return True
+    return False
+
+
+def format_plain_text(fn):
+    formatted = ''
+    with open(fn, 'r') as f:
+        for l in f:
+            l = l.rstrip()
+            if l.find('\t') != -1:
+                print(f'Warning: found tab in {fn}. Skipping...')
+                return
+            formatted += l + '\n'
+    while len(formatted) and formatted[-1] == '\n':
+        formatted = formatted[:-1]
+    formatted += '\n'
+    with open(fn, 'w') as f:
+        f.write(formatted)
+
+
+def _find_clang_format_bin():
+    candidates = ['clang-format-6.0', 'clang-format']
+    result = None
+
+    for c in candidates:
+        try:
+            if sp.run([c, '--version'], stdout=sp.DEVNULL,
+                      stderr=sp.DEVNULL).returncode == 0:
+                result = c
+                break
+        except:
+            pass
+    if result is None:
+        print(Fore.YELLOW +
+              'Did not find any clang-format executable, skipping C++ files',
+              file=sys.stderr)
+    else:
+        print('C++ formatter: {}{}'.format(Fore.GREEN, result))
+    print(Style.RESET_ALL)
+    return result
+
+
+def main(all=False, diff=None):
```

Comment:
BTW, can we also make `ti format` rely on this file's implementation, so that we don't have to repeat the code? (Not urgent)

Reply:
Will in a sep PR for all you want in #647 (also for `misc/build_doc.py` like this).